File: code/java/ch10/practice/Practice01Fibonacci.java
package ch10.practice;

import java.util.*;

/**
 * Practice 01: Fibonacci Number
 * ==============================
 * Chapter 10: The Magic of Recursion
 *
 * PROBLEM: Given a non-negative integer n, return the nth Fibonacci number.
 *          fib(0) = 0, fib(1) = 1, fib(n) = fib(n-1) + fib(n-2).
 *
 * ALGORITHM: Use recursion (memoization recommended for larger n).
 *
 * EXAMPLES:
 *   solve(0)  = 0
 *   solve(1)  = 1
 *   solve(10) = 55
 *   solve(15) = 610
 *
 * CONSTRAINTS: 0 <= n <= 30
 *
 * INSTRUCTIONS: Replace the body of solve() with your solution.
 */
public class Practice01Fibonacci {
    public static int solve(int n) {
        // TODO: Replace this with your solution
        return 0;
    }

    // ── Do not change anything below this line ──────────────────────
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = Integer.parseInt(sc.nextLine().trim());
        System.out.println(solve(n));
        sc.close();
    }
}

File: code/python/ch10/practice/practice_01_fibonacci.py
"""
Practice 1: Fibonacci
==============================
Chapter 10: The Magic of Recursion — Functions That Call Themselves

PROBLEM
-------
Return the nth Fibonacci number using recursion.
The Fibonacci sequence is defined as:
  F(0) = 0
  F(1) = 1
  F(n) = F(n-1) + F(n-2) for n >= 2

Plain recursion is O(2^n) which is very slow for large n.
Use memoization (a dictionary to cache results) to bring it down to O(n).

INPUT FORMAT
------------
A single integer n.

OUTPUT FORMAT
-------------
A single integer — the nth Fibonacci number.

CONSTRAINTS
-----------
- 0 <= n <= 50

EXAMPLES
--------
Input:
  0
Output: 0

Input:
  1
Output: 1

Input:
  10
Output: 55

HINT
----
Create a memo dictionary inside solve(). Write a helper function
fib(k) that checks the memo before computing, stores results in
the memo after computing, and uses F(0)=0, F(1)=1 as base cases.

INSTRUCTIONS
------------
Replace the `pass` in the solve() function with your solution.
The main block at the bottom handles input/output — don't change it.
"""


def solve(n: int) -> int:
    """Return the nth Fibonacci number using memoized recursion."""
    pass  # TODO: Replace this with your solution


# ── Do not change anything below this line ──────────────────────────
if __name__ == "__main__":
    n = int(input())
    print(solve(n))
